src/linker/Linker.Steps/ReflectionBlockedStep.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Collections.Generic;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+
+using System.Runtime.CompilerServices;
+
+namespace Mono.Linker.Steps {
+	public class ReflectionBlockedStep : BaseStep
+	{
+		public const string AttributeNamespace = "System.Runtime.CompilerServices";
+		public const string AttributeName = "DisablePrivateReflectionAttribute";
+
+		AssemblyDefinition assembly;
+		MethodReference noOptAttr;
+
+		static MethodDefinition _reflectionMethod;
+
+		public static MethodDefinition GetReflectionBlockedAttr (LinkContext context) {
+			if (_reflectionMethod != null)
+				return _reflectionMethod;
+
+			TypeDefinition methodImpl = BCL.FindPredefinedType(AttributeNamespace, AttributeName, context);
+			if (methodImpl == null)
+				throw new Exception("Could not find System.Runtime.CompilerServices.ReflectionBlockedAttribute in BCL.");
+
+			foreach (var ref_method in methodImpl.Methods)
+			{
+				if (!ref_method.IsConstructor)
+					continue;
+				if (ref_method.Parameters.Count != 0)
+					continue;
+				_reflectionMethod = ref_method;
+			}
+
+			return _reflectionMethod;
+		}
+
+		protected override void ProcessAssembly (AssemblyDefinition assembly)
+		{
+			if (Annotations.GetAction (assembly) != AssemblyAction.Link)
+				return;
+
+			this.assembly = assembly;
+
+			foreach (var type in assembly.MainModule.Types)
+				ProcessType (type);
+		}
+
+		void ProcessType (TypeDefinition type)
+		{
+			bool canAnnotateAll = true;
+			List<MethodDefinition> canAnnotate = new List<MethodDefinition> ();
+
+			foreach (var method in type.Methods) {
+				// Public methods have non-visible call sites by default, this attribute doesn't
+				// help us at all.
+				//
+				// See mono_aot_can_specialize in aot-compiler.c in mono
+				if (!method.IsPrivate)
+					continue;
+
+				if (!method.HasBody)
+					continue;
+
+				if (Annotations.HasUnseenCallers (method)) {
+					canAnnotateAll = false;
+					continue;
+				}
+
+				canAnnotate.Add (method);
+			}
+
+			if (canAnnotateAll) {
+				AnnotateType (type);
+			} else {
+				foreach (var method in canAnnotate)
+					AnnotateMethod (method);
+			}
+
+			foreach (var nested in type.NestedTypes)
+				ProcessType (nested);
+		}
+
+		void GetNoOptAttr ()
+		{
+			if (noOptAttr == null) {
+				noOptAttr = assembly.MainModule.ImportReference (GetReflectionBlockedAttr (Context));
```

Comment:
I thought you have to call `assembly.MainModule.ImportReference` once for each `assembly`?  This caching in `noOptAttr` looks like it would prevent that from happening.

Reply:
Ah, good catch. If this has to be regenerated once per assembly, I'll do that where I assign the assembly to the `assembly` object field. I'll push that change shortly.